### ISSUE TITLE
EME should support continuous key rotation per MPEG Common Encryption (ISO/IEC 23001-7

### ISSUE THREAD

**Issue by jdsmith3000** [w3c/encrypted-media#132]
The MPEG Common Encryption spec was specifically created to define “… standard encryption and key mapping methods that can be utilized to enable decryption of the same file using different Digital Rights Management (DRM) and key management systems”.  For key rotation, the 23001-7 protection system specific header box (‘pssh’) is expressly allowed to be in the movie fragment (‘moof’) box, enabling policy and key rotation scenarios, and “… MAY include data such as the URL of license server(s) or rights issuer(s) used, embedded licenses/rights, embedded keys(s), and/or other protection system specific metadata.” (23001-7, 8.1.1).  Keys distributed in the movie fragment are allowed so that they can be directly used by the client to decrypt and play content.  Cable and cellular networks need this key rotation ability to enforce re-authorization over multicast/broadcast. Millions of synchronized unicast license requests for multicast live broadcasts is inefficient and unnecessary. This is a primary reason why the ‘pssh’ box can be included in the movie fragment box.

EME currently doesn’t allow this rotation because it prohibits providing Initialization Data (‘pssh’ box) to the CDM except to generate a license request.  To support DRM-interoperability provided by 23001-7, EME must allow for initialization data to be provided to the CDM without generating a new session or performing a license acquisition, it must allow the CDM to derive a key from a ‘pssh’ box without requiring a key request message, and it must permit the ‘pssh’ box to contain a key.

This issue relates somewhat to #41 and #53, but recommends a specific use case we believe should be supported by the EME spec.

**Comment by greg80303** [comment]
This addresses some of the issues I brought up in #121 so I'm certainly in favor of this being addressed in the EME spec.

**Comment by adamgoldberg** [comment]
Continuous key rotation should be supported by EME, in favor of addressing this issue.

**Comment by pvanpeenen** [comment]
I'm also in favor of continuous key rotation support in EME

**Comment by chazlo** [comment]
Yes, I also agree with the proposed bug fix to EME to enable initialization data to be provided to the CDM without the latter having to generate a new session or perform a license acquisition, for key rotation support.

**Comment by steelejoe** [comment]
We have many customers who require this use case, so Adobe is in favor of this. I believe a solution to this would likely resolve the need issue 41 was intended to address.

**Comment by boblund** [comment]
We represent companies worldwide that plan to use EME and require this use case. CableLabs is in favor of this.

**Comment by andrejp** [comment]
Encrypted Media Extension API must support the ATSC, cable, MBMS and other use cases. Delivering hundreds keys is not efficient. I feel strongly that this architectural issue must be resolved prior to the first publication of EME.

**Comment by wilaw** [comment]
Akamai is in favor of this. Requiring millions of synchronized unicast license requests for live broadcasts will lead to failure-at-scale for both unicast and multicast delivered content.

**Comment by lpiron** [comment]
Nagra, as a worldwide provider of content protection solution, is also thinking that this needs to be address. Scalability has one of the key elements in Live television and not been efficient on this aspect is a showstopper.

**Comment by rchernock** [comment]
This capability is important for supporting next generation broadcast television (such as ATSC 3.0). I recommend acceptance of this bug to enable efficient broadcast key rotation.

**Comment by sandersaares** [comment]
I also consider it crucial that this feature gets proper support from EME at an early stage, as giving CDMs the necessary flexibility could be quite problematic to retrofit in a later version of EME.

**Comment by paulbrucecotton** [comment]
Given that we have multiple supporters for this "use case" (and so far no objectors), could @jdsmith3000 provide more information this week on how he proposes to implement this "use case" so that we might discuss this issue and those details at a Media TF meeting on Tue Dec 15?

**Comment by ddorwin** [comment]
@paulbrucecotton, while I encourage people to think about how such a feature might fit into EME and the web, we do object to focusing the group’s attention on it at this time. Details below.
<br/>

Thanks for all the comments. Congrats on making this the EME issue with the most unique participants. I positively appreciate the interest.

However, with no judgement on the feature, it is still a feature request. At this stage of EME - 4 years into the spec, when we finally have widely implemented, interoperable features - I am unequivocally opposed to feature requests distracting from and delaying stabilization and publishing of the spec.

Modifying the spec at the last minute to cram new features would be inconsiderate of the amount of time and effort expended to reach consensus, stabilize the spec, and address interoperability over the past 4 years as well as of the authors wanting to use the current baked feature set.

I also want to clarify a possible misunderstanding and suggest mechanisms to ensure that we move forward on this feature request.

EME v1 is _not the end of all EME features_. It is a base that can be extended in the future, as is common with web specs. The existing feature set has already enabled many providers to switch to HTML5, and the broad support and implementation are setting the stage for truly interoperable media and applications. Demonstrating that this group can effectively produce a Recommendation is a necessary condition for all future use cases. 

In the meantime, I suggest that those interested in such functionality incubate a solution that will see broad support from implementers and uses truly interoperable media (i.e. such that the feature does not rely on proprietary headers for each key system). Specific to this proposed solution, there have been no interoperable mechanisms even discussed, and that will be a prerequisite for inclusion in any version of the spec.

Finally, please note that this is **not** a judgement on the use case or specific solution. If this gets more content on the web in an interoperable way, I look forward to discussing mechanisms and methods to enable the wide reach that the web promises. However, we must not block v1’s path to CR and eventually REC.

**Comment by jdsmith3000** [comment]
We don’t view this as a feature request.  It’s a request to address a serious oversight in the EME spec against an original goal of supporting CENC.  Companies have been vocal with us about the issue, and it is one that we strongly believe must be addressed in V1.

There are two areas of oversight:

1.	Lack of support for inband key rotation (this issue #132)
2.	Difficulty in eliminating duplicate sessions/duplicate key requests (issue #121)

We believe both can be addressed with reasonable effort and without a major delay in the EME schedule.  We also don’t believe delays to the spec should be used though to justify postponing work on these problems.  

Specific changes we propose are:

1.	Allow simple inband keys to be directly consumed by the UA/CDM without triggering an `encrypted` event or license server messages.  The CENC specification intends for them to be directly used in the media pipeline, and we believe that should be honored in EME.  This alone could enable key rotation with minimal spec impact.
2.	Support a method on MediaKeys to test initData in the CDM with a response indicating whether a new license request should be processed.  This would be somewhat more work and agrees with some of the suggestions in issue #121.

I’ve posted two diagrams for discussion.  The first shows initData processing in EME as it is currently.

https://lists.w3.org/Archives/Public/www-archive/2015Dec/att-0005/EME_InitData_Current.pdf

The second shows it with “child keys” directly processed in the CDM without license server messages, and adds the initData check for duplicate keys/sessions

https://lists.w3.org/Archives/Public/www-archive/2015Dec/att-0006/EME_InitData_Proposed.pdf

**Comment by mwatson2** [comment]
Without commenting on the v1/v2 discussion, I'd like to address some technical points.

First, note that in the discussion of #121, an alternative characterization of the problem was in @greg80303 's statement: "I don't think the DASH manifest has enough information for me to be able to determine the disposition of each encrypted event.". That is not to say that we should not solve this problem in EME, but only to point out that it arises only because of the way DASH is constructed combined with the desire to build a generic DASH player that has no API other than the DASH manifest.

Second, I think we should have a logically solid conception of what a "session" is for in EME to guide our decisions on whether these more complex use-cases require multiple sessions. AFAIK, there are only two purposes for the session concept in EME:
1. As a context for message exchanges (that is, to allow multiple exchanges in parallel)
2. To define the granularity for persistence and removal of persisted data

Both of these are application-driven. At present, applications never need to create more than one session for a single playback unless they choose to for one of the above reasons. I think we should retain this principle that the purpose of sessions is for *application* control of message parallelism and granularity of persistence. We should avoid solutions which force multiple sessions for other reasons, as those will interfere with the above reasons.

As such, I'm not in favor of the proposed solutions which test whether keys referenced in mid-session `initData` are already present and force creation of a new session if they are not. For key rotation, as with key renewal, the message exchange should be in the context of the existing session unless the application thinks otherwise.

So, I think a better solution would be to allow provision of `initData` into an existing session at any time. This may or may not trigger a message exchange. This method would be very much like `update()`, except that it is `initData`, rather than message data, which is provided.

I am also not sure about the CDM directly processing in-band `initData`. At the start of a session, we deliberately provided support both for initData retrieved from the file and initData provided to the application by other means. It seems to me we should do the same here.

**Comment by jdsmith3000** [comment]
@mwatson2:  I generally agree with your stated reasons for EME sessions, and session re-use.  I'm not certain they address all cases where duplicate license requests may be issued, and that's part of what we believe should be corrected.  You appear to be endorsing a method on MediaKeySession though that might address this.  Do I understand that correctly?

We considered passing initData to the app and back to the CDM for child key processing, but the exchanges there did not seem to have a clear purpose other than to inform the app that an inband key was encountered.  On the downside, these keys may be frequently added to segments to support things like seeks, and so app traffic could be significant if it was involved in passing inband keys to the CDM.  It seems more natural and intended in the concepts from CENC that these inband keys would be directly consumed, as they are today in apps.

**Comment by mwatson2** [comment]
@jdsmith3000 Yes, I'm advocating a new method on MediaKeySession that would allow you to provide initData mid-session. This may or may not trigger a key exchange. (I'm not commenting on v2 vs v2 for this new method). This means that if the app does not want another session it does not need to create one.

I'll let @ddorwin comment on the reasons for passing initData via the application. I expect all the arguments that apply to initData in `moov` (it was a long debate) would also apply to initData in `moof`.

**Comment by nthorwir** [comment]
initData mid-session would mean there is a different initData category mid-seesion and start-session? 
I believe for simplicity, this decision should be in CDM not in application.

**Comment by mwatson2** [comment]
@nthorwir which decision ? The decision on whether or not to create a new session ? I think that should be for the application to decide, for the reasons in my post above. Introducing another purpose for multiple sessions - especially one that is not application driven - seems very problematic to me.

I don't see any reason for different initData 'categories', though I don't really know what you mean by that.

**Comment by nthorwir** [comment]
@mwatson2 yes, the decision to create a new session. 
I agree with the categories: i.e. I don't see a reason for different initData 'categories' either, but that's how I understood your suggestion to create 'a new method that would allow you to provide initData mid-session' (vs initData a the start of the session).

**Comment by mwatson2** [comment]
> @mwatson2 yes, the decision to create a new session. 

Ok, so could you address the points I made above arguing that multiple sessions should only be for application control of message parallelism and granularity of persistence ?
> I understood your suggestion to create 'a new method that would allow you to provide initData mid-session' (vs initData a the start of the session).

A green-field version of my proposal would have a single method to provide any kind of initData (`provideInitData()`, say). But it's been requested that we don't change existing methods lightly and given also that `generateKeyRequest()` is already in use in the field I would suggest a new method to provide initData mid-session. It doesn't mean the processing of the initData would be any different except that mid-session the initData may or may not trigger a message exchange.

**Comment by jdsmith3000** [comment]
@mwatson2,  I believe you are proposing that some `generateKeyRequest()` calls end without the key message exchanges.  This would accomplish passing initData to the CDM, and reflect a CDM decision that a new key is not needed.  Are there any situations where this decision should be under app control?

**Comment by mwatson2** [comment]
Let's step back (just a little): In order to support continuous key rotation there are (at least) three possible approaches:
(1) The application determines when a new key is needed and **opens a new session** for the new key. This can easily be accomplished by including information about needed keys in the common PSSH box.
(2) The application **opens a new session** for each received initData and the *CDM* determines whether a new key exchange is required
(3) No new session is opened - the initData is provided to an **existing session** and the CDM determines whether a key exchange is required

I am arguing for (3) on the basis that sessions are for application control of message parallelism and granularity of persistence only (and that adding additional purposes will interact badly with these existing purposes). We should not need to create a plethora of sessions for key rotation on a single stream.

The method used in (3) could be
(a) a new method for this purpose,
(b) a new method which *replaces* `generateKeyRequest()` and so serves to provide initData both at the beginning and during a session or
(c) the existing `generateKeyRequest()` method.

(c) would be confusing because it would not always generate a key request. It was requested that we don't break existing pages, which rules out (b). So I suggested (a).

In this context (that is, (3)+(a)), the issue of whether `generateKeyRequest()` always generates a key request is a *separate issue* more related to the idea of retrieving state from previous sessions by initData rather than by session id.

**Comment by jdsmith3000** [comment]
@mwatson2, I follow your thinking on session re-use, but not quite on whether `generateRequest()` always generates a key request.  My assumption is that would be specifically intended to avoid hitting license servers with duplicate requests for keys already available.  Data stored by sessionID doesn't seem useful for this.  The CDM can make this determination accurately though.

I think merging this test with `generateRequest` may well be a reasonable way to support this.  We proposed an initiData check method to preserve app control over whether a key request is processed or not.  That may not be required.

**Comment by mwatson2** [comment]
@jdsmith3000 One issue raised here and in #121 (and [bug 25268](https://www.w3.org/Bugs/Public/show_bug.cgi?id=25268)) is the case of initData arriving mid-stream and not knowing whether to establish a new session. The suggestion that `generateRequest()` not always generate a request has often been offered as a 'solution' to this problem. But I think that's a bad solution because you have to create a new session for every initData (see my previous comments). The mid-stream initData problem is better solved within a single session (as is suggest in the OP above).

Separately, we have other use-cases where multiple initData are available at the *start* of streaming (for example in different bitrate streams or for audio / video) and no other information is available about which streams share keys and which do not. If multiple message exchanges are in fact needed to get the necessary keys one would indeed want to do those in parallel (and thus need multiple sessions).

But, as per the discussion in #121 we should be sure that we understand the assumptions which lead us to this use-case before deciding on a solution and those have not yet been well-documented (I am still a bit lost as to how you can end up in a situation where the application does not know what keys it needs even at the beginning of the session).

Finally, it was suggested in previous debates on `generateRequest()` that a good way of retrieving persisted sessions would be by recognizing that a persisted session can fulfill the initData. We've discussed that before and concluded on retrieving persisted sessions by session id. I just wanted to clarify that this one is a separate issue, probably best dealt with separately.

**Comment by mavgit** [comment]
One more comment on requirements: We require key rotation for critical use cases, such as linear (live) TV and pay-per-view. Key rotation is supported and used today on websites using HTML4 plug-ins. Support for ISO CENC key rotation in EME is required to move these websites to HTML5 video.

**Comment by mwatson2** [comment]
FWIW, our use-cases wiki records both the method of updating an existing session with new keys and the method of creating a new session for key rotation as both "required" and "supported": https://www.w3.org/wiki/HTML/Media_Task_Force/EME_Use_Cases#License.2FKey_Rotation

**Comment by greg80303** [comment]
I think I agree with @mwatson2 that the application should be responsible for deciding when to create new sessions.  Whether or not initData associated with a key rotation triggers the creation of a new session is entirely up to the application.

@mwatson2 said:

> The suggestion that generateRequest() not always generate a request has often been offered
> as a 'solution' to this problem. But I think that's a bad solution because you have to create a new
> session for every initData

Are you saying that generateRequest() can only be called once per key session according to the spec?  I agree that this is the way I have typically used it, but is there something in the spec that specifically states this?

**Comment by mwatson2** [comment]
> Are you saying that generateRequest() can only be called once per key session according to the spec? 
> I agree that this is the way I have typically used it, but is there something in the spec that specifically 
> states this?

Yes, the first two steps in `generateRequest()`:
1. If this object's uninitialized value is false, return a promise rejected with a new DOMException whose name is InvalidStateError.
2. Let this object's uninitialized be false.

**Comment by mwatson2** [comment]
@ddorwin mentioned interoperability and I doubt we have consensus on what that means with respect to this feature.

There are several forms of interoperability we could consider. First, I expect we do have consensus on the idea that pages should not need browser- or keysystem-specific code. So, this feature should operate in the same way for all keysystems, as far as the page is concerned, and what this way is should be clear in the specification.

Second is media interoperability: it should be possible to use the same files with multiple keysystems. CENC itself addresses this with common encryption, but still requires keysystem-specific PSSH boxes, meaning that files must be updated to support a new keysystem. We addressed this last point for the simple streaming usecase by defining the [common PSSH format](https://w3c.github.io/encrypted-media/cenc-format.html#common-system}, support of which is **required** for all keysystems.

@ddorwin suggests above that a similar approach would be needed for key rotation. That is, a common header format for in-band keys. However, whilst it is technically obvious how this could be achieved, it would make me nervous unless the proposal came from one of the DRM vendors. It would be stepping into exactly the same territory as key delivery in key messages, which we have had to leave opaque.

It is also unclear that the same media interoperability concern applies to this use-case as applies to video-on-demand. In the VoD case, repackaging a large library to include a new DRM header is a major undertaking (we have done it several times at considerable cost). For live services, there is no stream before the live event itself, so the requirement to add a new header is of a different character (upgrading software).

I wonder if the proponents of this feature could elaborate on the characteristics of the applications for key rotation and whether media interoperability as discussed above is a concern ?

**Comment by joeyparrish** [comment]
With regard to repackaging a large VOD library, this can be avoided if you are using DASH, as you can add PSSHs to the manifest that override what's in the content.  Instead of repackaging to add a key system, you just have to add an additional PSSH element to each manifest.

**Comment by mwatson2** [comment]
> With regard to repackaging a large VOD library, this can be avoided if you are using DASH, as you can
> add PSSHs to the manifest that override what's in the content. Instead of repackaging to add a key
> system, you just have to add an additional PSSH element to each manifest.

For sure. Is that solution application to the key rotation use-case ?

**Comment by joeyparrish** [comment]
>> With regard to repackaging a large VOD library, this can be avoided if you are using DASH, as you can add PSSHs to the manifest that override what's in the content. Instead of repackaging to add a key
system, you just have to add an additional PSSH element to each manifest.

> For sure. Is that solution application to the key rotation use-case ?

Not as far as I know, but I was under the impression that key rotation was typically used in live streams, not VOD.

**Comment by nthorwir** [comment]
Key rotation is used for live streams and becomes relevant in VOD when  live stream is recorded for later distribution (catch up TV). 

The pssh in manifest is helpful to prevent media update but I believe it requires a new DASH Period for each new key interval and that is not desirable : 1. period boundaries should be determined by content requirements not content protection requirements and 2. a period switch triggers re-initialization in some environments that is not seamless  in playback. The media update to add content protection systems is undesirable but may be acceptable as it should be infrequent. 
 
I believe, one main issue of key rotation is how the keys are handled efficiently and without a client storm. Also flagged in the use cases https://www.w3.org/wiki/HTML/Media_Task_Force/EME_Use_Cases#License.2FKey_Rotation as a related bug  (https://www.w3.org/wiki/HTML/Media_Task_Force/EME_Use_Cases#License.2FKey_Rotation).
There are at least 2 solutions in use: 
a) Key hierarchy: an individual keys is delivered to each client and control messages with a leaf key enable decryption after that. 
b) Key pre-fetch: future KIDs are delivered to CDMs. CDM picks a random time to fetch the key, thereby distributing key request evenly.
They have different properties but require communication within the pssh in a specific way. 

I don't understand the common system requirement https://w3c.github.io/encrypted-media/cenc-format.html#common-system. It seems that the protection system specific header (pssh) is used as a protection system common header. I can see the use to make the KID public, e.g. using pssh ver 1, but believe the pssh should be specific to allow use cases such as efficient key rotation.

**Comment by mwatson2** [comment]
> I don't understand the common system requirement https://w3c.github.io/encrypted-media/cenc-
> format.html#common-system. It seems that the protection system specific header (pssh) is used as a 
> protection system common header. I can see the use to make the KID public, e.g. using pssh ver 1, but > believe the pssh should be specific to allow use cases such as efficient key rotation.

The idea is exactly that by using the common header (only) you do not need to have key-system-specific files, at least for the use-cases that have received active consideration before this issue.

@ddorwin suggests that a similar solution is required for key rotation, but I think at least for live the considerations are not the same: the requirement for key-system-specific material in the live media with key rotation is reasonable (as well as unavoidable, IMO).

For VOD of previously live material, your solution (b) allows those files to be played by a new DRM without modification of the files, although that new DRM may be disadvantaged compared to those whose keysystem-specific material is embedded in the media.

**Comment by joeyparrish** [comment]
> The pssh in manifest is helpful to prevent media update but I believe it requires a new DASH Period for each new key interval and that is not desirable

I agree that it would not be desirable if a new Period were required, but it is not true for the live DASH streams I've seen with key rotation.  One Period, PSSHs (and therefore 'encrypted' events) from non-initialization segments.  With current EME APIs, Shaka Player creates a new session for each new key and monitors session expiration to clean up and close sessions that are no longer needed.

**Comment by mavgit** [comment]
From the first EME draft v0.1, authored by Dorwin, Bateman, & Watson, key rotation is explicitly supported:
> **Is key rotation supported?**
>
> Yes.

https://dvcs.w3.org/hg/html-media/raw-file/eme-v0.1/encrypted-media/encrypted-media.html#faq-use-cases

I'm not aware of any consensus decision to remove this use case, which is why I consider key rotation to be still supported.

**Comment by mwatson2** [comment]
@mavgit Yes, that text indicates that we intended to cover this use case. Our [use case wiki](https://www.w3.org/wiki/HTML/Media_Task_Force/EME_Use_Cases) also indicates that it is supported, but describes it in terms of providing new initData and creating new sessions (which *is* supported in the spec).

If our intention was to address key rotation in a broader sense, then what has happened is that we have not made good on our intention. We should certainly do so now and I don't think there is any disagreement on that point.

What is at issue, is whether we should include that in v1.0. As things stand, I believe everything in the specification has multiple implementations and so we could proceed through to CR quite rapidly. If we introduce material which is not yet implemented, then things will be slower.

I believe after v1.0 we can more to a more rapid publication cycle and in fact I think that anything in the v1.0 Last Call which does not already have multiple implementations should be moved out into v1.1, rather than waiting for implementation.

**Comment by steelejoe** [comment]
@mwatson2 The currently support version of key rotation would be more accurate to call *license rotation* given its described workflow. My concern with that approach is that for customer with linear/live feeds today using key rotation, the period of rotation may be small enough that the performance will dramatically suffer. E.g. if the keys rotate every 2 min across a large population of simultaneous viewers, the resulting key server load may be unsustainable. I will leave it to the operators on the thread to talk about what average key periods are, but I will say that a mechanism using keys already acquired to decrypt keys found in initData can handle this easily.

**Comment by mwatson2** [comment]
@steelejoe I completely agree.

**Comment by jdsmith3000** [comment]
An updated list that removes editor and chairman names is below.

GitHub has additional comments from from these folks.  Are any recognized?  Any I've missed?

* Squelart
* Greg80303
* Mavgit
* Adamgoldberg
* Pvanpeenen
* Chazlo
* Andrejp
* Wilaw
* Ipiron
* Rchernock
* Sandersaares
* Nthorwir
* Joeyparrish

>Acknowledgments

>The Media Task Force chairman and editors would like to thank the people listed below for their useful comments and contributions, both large and small, that have led to changes to this specification over the years.

>Contributors:

>Aaron Colwell, Alex Giladi, Alex M (Coyo), Andreas Kuckartz, Andrew Livingston, Anne van Kesteren, Árni Arent, Arthur Barstow, Aymeric Vitte, B. Ross Ashley, Benjamin Rainer, Bjoern Hoehrmann, Bob Lund, Boris Zbarsky, Brad Hill, Brendan Aragorn, Brian Evans, Brian Smith, Casey Callaghan, Catherine Roy, Charles McCathie Nevile, Michael Chen, Chris Pearce, Christine Runnegar, Cyril Concolato, Dan Dart, Daniel Appelquist, Danny O'Brien, David LaPalomento, David Singer, Deivi Kuhn, Domenic Denicola, Doug Schepers, Duncan Rowden, Emmanuel Poitier, Emmanuel Revah, Eric Carlson, Florian Bösch, François Remy, Fred Andrews, Gaël Pegliasco, Gary Hughes, Gilles Boccon-Gibod, Giuseppe Pascale, Glenn Adams, Guy Paskar, Weiss Hadar, Hans Schmucker, Henri Sivonen, Bin Hu, Hugo Bodewig, Ian Hickson, Tatsuya Igarashi,  James Craig, James Graham, Janina Sajka, János Barta, Jay Munro, Jer Noble, Jerry Smith, Jim Ley, Joe Feely, Joe Steele, Joey Parrish, John C. Vernaleo, John Foliot, John Kemp, John Simmons, John Sullivan, Joseph Ashwood, Joseph Lorenzo Hall, Julio Serrano, Kevin Kershaw, Kevin Streeter, Kornel Lesiński, Leif Halvard Silli, Jason Lewis, Maciej Stachowiak, Antoine Maillard, Marcos Caceres, Mark Nottingham, Martin Soukup, Shinya Maruyama, Matt Field, Matt Matrisciano, Matt Womer, Matthew Gregan, David Mays, Michael Aubertin, Michael Hausenblas, Michael Thornburgh, Michael Smith, Muñiz Piniella, Niels Thorwirth, Pascual Conesa, Petr Peterka, Philip Jägenstedt, Philippe Le Hegaret, Pierre-Anthony Lemieux, Raphaël Troncy, Rob Manson, Robert O'Callahan, Robin Berjon, Robin Boast, Romain Bouqueau, Ryan Sleevi, Sam Ruby, Silvia Pfeiffer, Simon Warner, Stefan Lederer, Steve Faulkner, Steve Heffernan, Steven Robertson, Stig Andersen, Eric Sunyang, Tab Atkins Jr., Takeshi Yoshino, Tobie Langel, Mark Vickers, Wayne Borean, Wendy Seltzer, Yang Sun, Zak Fenton, and Zhouhaojun.

**Comment by greg80303** [comment]
greg80303 = Greg Rutz (CableLabs)

**Comment by sandersaares** [comment]
I do not quite follow what the context of this last comment is but if you are asking for a mapping from GitHub username to real name, sandersaares = Sander Saares.

**Comment by boblund** [comment]
Boblund is missing and = Bob Lund
________________________________

**Comment by adamgoldberg** [comment]
And ‘adamgoldberg’ = Adam Goldberg – and I’m the Chair of ATSC TG-3/S36


Adam Goldberg
AGP, LLC
+1-202-507-9900

**Comment by adamgoldberg** [comment]
Oh, and while I’m at it:

* Pvanpeenen  = Pete Van Peenen

* Rchernock = Rich Chernock




Adam Goldberg
AGP, LLC
+1-202-507-9900

From: Adam Goldberg [mailto:adam@agp-llc.com]
Sent: Friday, January 22, 2016 11:54 AM
To: w3c/encrypted-media <reply@reply.github.com>; w3c/encrypted-media <encrypted-media@noreply.github.com>
Cc: adamgoldberg <adam_g@yahoo.com>
Subject: RE: [encrypted-media] EME should support continuous key rotation per MPEG Common Encryption (ISO/IEC 23001-7) (#132)

And ‘adamgoldberg’ = Adam Goldberg – and I’m the Chair of ATSC TG-3/S36


Adam Goldberg
AGP, LLC
+1-202-507-9900

**Comment by mwatson2** [comment]
I think these most recent comments are intended for Issue #109.

**Comment by lpiron** [comment]
and lpiron = Laurent Piron

Regards

**Comment by mavgit** [comment]
mavgit is me.

On Jan 22, 2016, at 10:03 AM, lpiron <notifications@github.com<mailto:notifications@github.com>> wrote:

and lpiron = Laurent Piron

Regards

**Comment by chazlo** [comment]
Chazlo = Charles Lo

Cheers

**Comment by jdsmith3000** [comment]
Correct.  This discussion belongs on #109, but at least here it answered many of my alias questions (by accident).  I will update the draft and move the conversation to #109.

**Comment by ddorwin** [comment]
Triaging per https://lists.w3.org/Archives/Public/public-html-media/2016Mar/0003.html.

**Comment by mavgit** [comment]
I strongly object to making this VNext. We already have two or more implementations that support this feature. It is not required to have support from 100% of implementations - just 2. We have had more voiced support for this feature than any other.

We just need to agree on how to allow it in the spec. There are several options for how to support this, some requiring little change to the spec. I'd like to discuss this on a call and agree on the best support for V1.

**Comment by mwatson2** [comment]
In my opinion, the simplest way to address this issue is to provide mid-stream initData (received in a `needskey` event) using the `update()` method. This may not require any changes to the specification.

The procedures for `update()` allow that processing the message may or may not result in addition of keys to the session and may or may not result in a message exchange, so this would seem to cover the various use-cases that have been discussed. There is nothing which prevents an unsolicited `update()` call.

This method has the advantage that we do not need to make any changes to the data flow architecture (initData is still always provided via the application) and that the application remains in control of session granularity (see [here](https://github.com/w3c/encrypted-media/issues/132#issuecomment-163793743) for why that is desirable).

There is a minor terminology issues, since the `update()` procedures call the message a "response".

Of course, this requires that DRMs can disambiguate initData from other kinds of DRM message, but that does not seem much of a challenge.

If this is acceptable, I will update the [use-case](https://www.w3.org/wiki/HTML/Media_Task_Force/EME_Use_Cases#License.2FKey_Rotation) to describe this approach in more detail and perhaps we should reclassify this issue as `V1NonBlocking` for the terminology issue ?

**Comment by greg80303** [comment]
@mwatson2 How will the application know whether the `needkey` initData should be passed to `update()` or to `generateRequest()`?

**Comment by mwatson2** [comment]
That's up to the application and it's free to choose for its own reasons. As noted (here)[https://github.com/w3c/encrypted-media/issues/132#issuecomment-163793743], the purpose of sessions is (a) to enable multiple license requests to proceed in parallel and (b) to define the granularity of the record of key usage or proof or license destruction. If you don't need either of these things you would never need more than one session, otherwise the creation or not of new sessions should be based on those two factors.

**Comment by joeyparrish** [comment]
@mwatson2, what you're describing seems to have two modes of operation.  One is to use initData to update() an existing session, the other is to generateRequest() on a brand new session.  An application that does not know about the former may fall back to the latter.

Since new sessions & generateRequest already allows key rotation today in existing implementations, without changes, why do we need to allow passing initData to update() at all in V1?

**Comment by joeyparrish** [comment]
Just to clarify my previous comment, I don't disagree with passing initData to update() in principle, I just don't understand why it's critical enough to push into V1 when there's already a perfectly good key rotation mechanism in use today.

**Comment by mwatson2** [comment]
@joeparrish The model where a new key exchange (with the server) is needed for each new key has more in common with "key renewal", not key rotation. IMO this should be triggered when needed within a single session by the DRM, not by the provision of initData.

If there are use-cases where each key exchange needs to be informed by a new initData, then I think the present approach of requiring a new session for each key exchange is a deficiency: you should not need a new session for the new exchange.

IIUC, this issue is asking for a mechanism that would support key chaining, where the new key is delivered in-band *without server interaction*. For this to work it is necessary for the initData to be passed to the CDM. Creating a new session for this is obviously wrong, since there is no requirement for server interaction. Passing the initData directly to the CDM (rather than via the application) does not align with our current data flow model (decided after long discussion).

Obviously, one would not be able to play content that relied on such key chaining with an application or CDM that did not support it.

Furthermore, there has been discussion of applications which are unaware of when a new key is required and when it is not. Whilst I have trouble understanding why an application would be so unaware, it also seems reasonable that an application could simply pass down all the initData it receives to the CDM and let the CDM decide when a message exchange is required.

I'm not arguing for anything to be "pushed into V1". Others have asked for this issue to be addressed in V1. I'm pointing out a way that it can be addressed without specification changes. That is, I am saying that these use-cases are all already supported (in the way I have described) and that this is a non-issue.

**Comment by steelejoe** [comment]
@mwatson2 I agree with this proposal. It will allow applications to reduce the number of sessions and server interactions required where the CDM supports it and has a clear fallback when the CDM does not. It also preserves the behavior for any existing applications and it does not introduce any new APIs. Adobe's CDM implementation could support this.

**Comment by mavgit** [comment]
@mwatson2 - I support this proposal. It will enable us to continue to deliver our linear (live) TV channels in addition to our on-demand video, as we do today using Flash. Without this fix, we can only deliver on-demand video, which would be a significant decrease in functionality from our current website. This proposal also increases EME compatibility with MPEG Common Encryption.

While in general, I'm neutral on what syntax is used to support a given functionality, this proposal is especially  compelling at this time because it has little or no change to the EME API set. There are also multiple CDM/DRMs that are integrated with EME that already support this functionality.

**Comment by ddorwin** [comment]
I was asked to respond to the specific proposal from @mwatson2. So, for the record:

We object to actions that enable interop problems by not clearly specifying required behavior. Our overall position remains the same:
* We object to “solving” continuous key rotation in V1 as it is underspecified, not interoperable at this time, and cannot be appropriately addressed while adhering to the current timeline.
* We strongly urge those interested in such key rotation to incubate their solution, specify the use cases, and continue to work on an interoperable solution for vNext

<br/>While the latest proposal from @mwatson2 may not require spec changes, it promotes interop problems by not clearly specifying required behavior. We are very interested in working on interoperable solutions that help bring such content to the web so that all users can access it. We urge proponents to incubate the solution and move forward on an interoperable solution.

**Comment by mwatson2** [comment]
Clearly, there is market demand for this feature and implementation is likely to go ahead with or without action in this group. Such implementation work is exactly the kind of experimentation that is needed to drive towards an interoperable solution.

I do feel it would be valuable to provide some direction from this group and my [proposal](https://github.com/w3c/encrypted-media/issues/132#issuecomment-195427458) above seems to have some support. Other suggested solutions (such as passing initData directly to the CDM) have problems.

So, how about the following:
- raise a new V1NonBlocking bug to clarify the text and / or add a note indicating that data provided in update() calls is entirely CDM-specific (presently the text describes it as a "message" or "response")
- move this bug to VNext with the intention of standardizing an interoperable version of this feature after a period of incubation / experimentation
- update the Wiki to describe the proposed use of update() as a possible approach for this use-case

**Comment by mavgit** [comment]
@ddorwin wrote: "While the latest proposal from @mwatson2 may not require spec changes, it promotes interop problems by not clearly specifying required behavior."

What behavior isn't specified? Can you be specific about what needs to be improved? Can you be specific about what problems are promoted?

@ddorwin wrote: "We urge proponents to incubate the solution and move forward on an interoperable solution."

These ideas are beyond incubation. Key rotation was born years ago and has been running under Flash and Silverlight for years. Key rotation also runs under multiple EME CDMs today and has done so for some time. @mwatson2 's proposal is just the latest proposal. We just need to have all authors willing to join in the conversation about this, as we have on every other bug except this.

**Comment by ddorwin** [comment]
@mavgit wrote:
>What behavior isn't specified? Can you be specific about what needs to be improved? Can you be specific about what problems are promoted?

There is no documentation of the behavior an implementation would need to implement in order to play such content nor is there a definition of interoperable content. With neither of these documented, implying that the content is supported by EME would encourage non-interoperable content and segmentation of the platform.

>These ideas are beyond incubation. Key rotation was born years ago and has been running under Flash and Silverlight for years.

It's not _key rotation_ that needs incubation - it is how to support these use cases, which may include continuous key rotation, in the web platform.

DRM has also existed for many years, but it has been quite an effort to enable it in the web platform, including enabling/encouraging interoperability, addressing security and privacy concerns, etc. The same is true of Bluetooth, USB, and many other capabilities. (Incubation is a tool that did not exist when EME started so a lot of investigation and experimentation happened in the WG. Today, incubation is an important part of the spec process.)

>Key rotation also runs under multiple EME CDMs today and has done so for some time.

Most of the proposals in this issue involve participation from the application, and I haven't heard anyone say that such solutions have been implemented. I don't believe whatever CDMs might be doing today is a good indicator of maturity.

>We just need to have all authors willing to join in the conversation about this, as we have on every other bug except this.

Incubation is the appropriate process for conversations about new features, especially when there is no clear solution as in this case. Most importantly, V1 is close to completion and must not be held up while we incubate and/or experiment with new functionality. This is not the only issue that has been deferred to incubation for [VNext](https://github.com/w3c/encrypted-media/milestones/VNext).

**Comment by paulbrucecotton** [comment]
@steelejoe, @mwatson2, @jdsmith3000, @mavgit, @joeyparrish, @greg80303, @nthorwir and @ddorwin 

This issue is currently marked as VNext.  See my response to Joe Steele about the implications of this (repeated below):
https://lists.w3.org/Archives/Public/public-html-media/2016May/0049.html

> However some EME issues marked as V.Next (e.g. issue 132) are still under discussion as to the appropriate milestone. This seems like it would effectively shut down that discussion. Am I interpreting that correctly or would that still be an ongoing discussion?

The WG can continue to discuss VNext issues like [1] but after June 9 we will NOT be able to make any "substantive changes" to the EME specification for this or any other issues in order to stay on our published timeline.

If you or other WG members believe that changes need to be made to the EME specification to either:
a) clarify existing support for this feature or
b) to explicitly add support for this feature then you need to make any such proposals ASAP.

For example, I invite you to review Mark Watson's comment at [2] in which Mark proposes to split out VINonBlocking issues to do a).

/paulc

[1] https://github.com/w3c/encrypted-media/issues/132
[2] https://github.com/w3c/encrypted-media/issues/132#issuecomment-208981634

**Comment by xhwang-chromium** [comment]
Reviving this old issue (~9 years) which aims to add key rotation support to EME. It's an important feature with many users and supporters back then. Today key rotation is being used for live stream, e.g. on TV platforms. It'd be great if we can add native support of this feature for better interoperability.

For anyone interested, I strongly recommend you to read this (long) thread, which has a lot of history and information. There were some good discussions on how to support it, but back then there were some pressure to finalize EME V1 and this issue was deferred to incubation for VNext. This was punted again during EME V2 discussions.

Recently this issue caught my attention, and I also found that some applications and CDMs are using some non-standard way to use EME to support key rotation, e.g. to use `generateRequest()` to pass in the `initData` that contains the license/key, and expects an empty [MediaKeyMessage](https://www.w3.org/TR/encrypted-media/#dom-mediakeymessageevent) to indicate that no requests need to be sent to the server.

Given the importance of the use case, I feel it's worth a discussion to see whether we can find a way to support this feature in EME natively.

**Comment by cconcolato** [comment]
Are there any public bitstreams that demonstrate how key rotation are used in real world deployments?

**Comment by xhwang-chromium** [comment]
> Are there any public bitstreams that demonstrate how key rotation are used in real world deployments?

There are test streams using key rotation, but I am not sure whether a public one is available. I'll find out and update back here.

**Comment by xhwang-chromium** [comment]
For people not familiar with continuous key rotation, here's my understanding on how the system works. And Microsoft [PlayReady License Chaining](https://learn.microsoft.com/en-us/playready/overview/license-chaining) is such an example. Google Widevine has a similar feature. The following is my high level understanding, which might not be accurate in details:
- The first init data contains information for the CDM to generate a request to fetch a license containing a "root" key.
- The subsequent init data (embedded in the content) contains some form of information containing encrypted content key, which can be decrypted using the "root" key.
- The "content key" changes from section to section, or across different titles in one playback group, which is why it's called "key rotation".
- Since the subsequent init data is embedded in the content, and the CDM already has the "root" key, no license exchange is needed to continue the playback from section to section, or to play all titles in the same playback group.

**Comment by xhwang-chromium** [comment]
To quote the requirement from the initial post:
> To support DRM-interoperability provided by 23001-7, EME must allow for initialization data to be provided to the CDM without generating a new session or performing a license acquisition, it must allow the CDM to derive a key from a ‘pssh’ box without requiring a key request message, and it must permit the ‘pssh’ box to contain a key.

Previous discussions are focused on:
- Whether to use a single session or multiple sessions to manage the initial and subsequent init data.
- How to pass the subsequent init data to the CDM, using `generateRequest()`, `update()`, or a new method?

In reality, I've seen systems where all the root/content keys are managed together in one session, e.g. follow the same policy, and systems where the root/content keys are managed separately, e.g. having different policies. For minimum change of the the EME spec, and maximum compatibility with existing JavaScript applications and CDM implementations, I tentatively propose the following changes to the EME spec for discussion:
- `generateRequest()` is not required to always generate a request.
- Add a new [MediaKeySessionClosedReason](https://www.w3.org/TR/encrypted-media-2/#dom-mediakeysessionclosedreason) enum value: “initdata-processed”. (Exactly naming TBD)

This should support both the single and multi session models. The application will always create a new session on a new init data, and the CDM can decide whether the new session can be merged into an existing session (by closing the session with "initdata-processed") to keep one single session, or to keep the new session alive to have multiple sessions.

Single session case:
![5mAU3w6Rq3XkXmn](https://github.com/user-attachments/assets/80721e3c-d0b8-47c6-bca2-a0a208184f9f)

Multi-session case:
![8R5x2EM3rsrAkRN](https://github.com/user-attachments/assets/43e194e4-ad76-4539-8f84-052b2a50de28)

**Comment by chrisn** [comment]
Discussed at TPAC 2024: [minutes](https://www.w3.org/2024/09/26-mediawg-minutes.html#nt03)

**Comment by chrisn** [comment]
@xhwang-chromium At TPAC 2024 we decided the next step was to write an explainer for this feature. Should we plan to do this and discuss at TPAC 2025?

**Comment by xhwang-chromium** [comment]
Yes, will do and that's the plan. Sorry for the delay!

**Comment by xhwang-chromium** [comment]
Explainer for discussion: https://github.com/w3c/encrypted-media/blob/main/key-rotation.md
